### PR TITLE
Removal of unused `signed_struct` to reduce OpenSSL requirement

### LINKED ISSUE #86: Add version information to signed structs.
Currently signed structs produce JSON with no version information. We should add a version indicator to indicate what version of a signed struct type a JSON document is generated from.

### PR DESCRIPTION
**Pull request by AbhijithGanesh**
Signed-off-by: Abhijith Ganesh <67182544+AbhijithGanesh@users.noreply.github.com>

## PR Checklist

- [x] I've built the code `cargo build`. For major changes, `cargo build --workspace --release` is recommended.
- [x] I've run the automated unit tests `cargo test`. This executes our automated unit tests.
- [x] I've not broken any existing tests or functionality. In addition to `cargo test`, I've run `cargo clippy`
- [x] I've not introduced any new known security vulnerabilities. I've run `cargo audit`

- [x] I've included a brief description and a good title of the proposed changes and how to test them.
- [x] I've read the [contributing guidelines](https://github.com/pyrsia/.github/blob/main/contributing.md).
- [x] I've associated an [issue](https://github.com/pyrsia/pyrsia/issues) with this Pull Request. 
- [x] I've checked that the code is up-to-date with the `pyrsia/main` branch.
- [x] I've read ["What is a Good PR?"](https://github.com/pyrsia/pyrsia/blob/main/docs/good_pr.md)
- [x] I've assigned this Pull Request to "pyrsia/collaborators"

## Description

Fixes pyrsia/pyrsia#484 
Closes #86 

This PR deletes signed_struct, signed effectively removing the OpenSSL dependency for the Node.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -46,6 +43,7 @@ itertools = "0.10.3"
 lava_torrent = "0.5.0"
 lazy_static = "1.4.0"
 libp2p = { version = "0.43.0", features=["tcp-tokio"]}
+libp2p-kad = "0.35.0"
```
Comment (prince-chrismc) [inline Cargo.toml:46]: Looks like a merge conflict

File: src/docker/v2/handlers/manifests.rs [hunk 2]
```diff
@@ -454,13 +448,17 @@ fn add_fslayers(artifacts: &mut Vec<Artifact>, fslayer: &Value) -> Result<(), an
...
+    artifacts.push(Artifact {
+        algorithm: HashAlgorithm::SHA256,
+        hash: digest,
+        mime_type: Some(MEDIA_TYPE_BLOB_GZIPPED.to_string()),
+        creation_time: Some(String::from("")),
+        source_url: Some(String::from("")),
+        size: Some(12345678),
+        url: Some(String::from("")),
+        name: Some(String::from("")),
+        metadata: Map::new(),
```
Comment (prince-chrismc) [inline src/docker/v2/handlers/manifests.rs:460]: this should probably be `None` since they were previously unset

File: src/docker/v2/handlers/manifests.rs [hunk 3]
```diff
@@ -413,14 +405,17 @@ fn package_version_from_schema1(
...
+        creation_time: Some(String::from("")),
+        source_url: Some(String::from("")),
+        algorithm: hash_algorithm,
+        hash: hash,
+        mime_type: Some(MEDIA_TYPE_SCHEMA_1.to_string()),
+        size: Some(size64),
+    });
```
Comment (prince-chrismc) [inline src/docker/v2/handlers/manifests.rs:418]: there might be more `None`s here

File: src/metadata_manager/metadata.rs [hunk 4]
```diff
@@ -358,24 +327,18 @@ mod tests {
...
-            JwsSignatureAlgorithms::RS512,
-            &key_pair.private_key,
-            &key_pair.public_key,
-        )?;
+        let namespace_clone = path.clone();
+        let namespace = Namespace {
+            id: id,
+            package_type: PackageTypeName::Docker,
+            namespace_path: namespace_clone,
+            administrators: vec![vec![]],
```
Comment (prince-chrismc) [inline src/metadata_manager/metadata.rs:338]: Maybe `None`

File: src/node_manager/handlers.rs [hunk 5]
```diff
@@ -207,12 +207,12 @@ mod tests {
         assert_eq!(s, s1);
     }
 
-    #[test]
-    fn test_that_a_metadata_manager_is_created_and_accessible() {
-        let untrusted_key_pair = METADATA_MGR.untrusted_key_pair();
-        assert!(!untrusted_key_pair.public_key.is_empty());
-        assert!(!untrusted_key_pair.private_key.is_empty());
-    }
+    // #[test]
```
Comment (prince-chrismc) [inline src/node_manager/handlers.rs:210]: please delete

File: src/metadata_manager/metadata.rs (hunk 4 above)
Comment (AbhijithGanesh) [reply src/metadata_manager/metadata.rs:338]: Nope, it's more than that

File: src/docker/v2/handlers/manifests.rs (hunk 3 above)
Comment (AbhijithGanesh) [reply src/docker/v2/handlers/manifests.rs:418]: The exact problem with that concept is that, if I go ahead and make 
`
Size: None
`
 It breaks with the test case at line 917

Comment (AbhijithGanesh) [reply src/docker/v2/handlers/manifests.rs:418]: This is a bit of hard situation

Comment (AbhijithGanesh) [reply src/docker/v2/handlers/manifests.rs:418]: I can not change size alone and it should work

Comment (prince-chrismc) [reply src/docker/v2/handlers/manifests.rs:418]: there might, there was an extra feild

File: src/node_manager/model/artifact.rs [hunk 6]
```diff
@@ -27,30 +27,30 @@ use serde_json::{Map, Value};
 #[derive(Builder, Debug, Serialize, Deserialize, Clone, PartialEq)]
 pub struct Artifact {
     /// The hash value that identifies the artifact.
-    hash: Vec<u8>,
+    pub hash: Vec<u8>,
```
Comment (fishseabowl) [inline src/node_manager/model/artifact.rs:30]: Why expose private fields?

File: src/node_manager/model/artifact.rs [hunk 7]
```diff
@@ -98,11 +98,17 @@ mod tests {
...
+            creation_time: Some(String::from("")),
+            source_url: Some(String::from("")),
+            url: Some(String::from("")),
+            hash: vec![0x38u8, 0x4fu8],
+            algorithm: HashAlgorithm::SHA256,
+            metadata: Map::new(),
+            size: Some(60),
+            mime_type: Some(String::new()),
+            name: Some(String::from("acme")),
+        };
```
Comment (fishseabowl) [inline src/node_manager/model/artifact.rs:111]: It may be better to add a init/default method

Comment (AbhijithGanesh) [reply src/node_manager/model/artifact.rs:111]: It is, I was under the impression that ArtifactBuilder was stripped, now that I know it exists, I shall revert Artifact

File: src/node_manager/model/artifact.rs (hunk 6 above)
Comment (AbhijithGanesh) [reply src/node_manager/model/artifact.rs:30]: Apologies, since I was using non-builder method, I couldn't access them. I will be reverting immediately

File: src/node_manager/handlers.rs (hunk 5 above)
Comment (AbhijithGanesh) [reply src/node_manager/handlers.rs]: consider it done.

**Comment by tiainen** [comment]
Just so that I understand correctly, this PR will cause all metadata to be stored in plain unsigned JSON?

File: Dockerfile [hunk 8]
```diff
@@ -40,7 +40,6 @@ RUN <<EOT bash
     apt-get update
     apt-get install -y \
         ca-certificates \
```
[the remark this answers is not in the document]
Comment (AbhijithGanesh) [reply Dockerfile:42]: I wanna know why though? Its just a  newline.

**Comment by AbhijithGanesh** [comment]
For the time being, the idea of signing will be re-implemented, till then unsigned metadata

File: Dockerfile (hunk 8 above)
[the remark this answers is not in the document]
Comment (prince-chrismc) [reply Dockerfile:42]: it would awkwards run both statements in the same line... harder to see it fail

File: Cargo.toml (hunk 1 above)
Comment (prince-chrismc) [reply Cargo.toml:46]: please delete this it was removed in #483

File: Cargo.toml [hunk 9]
```diff
@@ -63,6 +61,7 @@ strum_macros = "0.24.0"
 tempfile = "3.2.0"
 test-log = "0.2.8"
 thiserror = "1.0.30"
+time = { version = "0.3.6", features = ["formatting", "parsing"] }
```
Comment (prince-chrismc) [inline Cargo.toml:64]: I was not expecting this 🤔 why is this required?

File: src/docker/v2/handlers/manifests.rs [hunk 10]
```diff
@@ -512,7 +503,7 @@ fn package_version_from_manifest_list(
     docker_name: &str,
     docker_reference: &str,
     hash_algorithm: HashAlgorithm,
-    hash: Vec<u8>,
+    _hash: Vec<u8>,
```
Comment (prince-chrismc) [inline src/docker/v2/handlers/manifests.rs:506]: this looks out of place since it's being used below

File: src/metadata_manager/metadata.rs [hunk 11]
```diff
@@ -15,17 +15,17 @@
 */
 extern crate anyhow;
 
+use super::model::artifact::ArtifactBuilder;
```
Comment (prince-chrismc) [inline src/metadata_manager/metadata.rs:18]: clippy warning

```suggestion
```

File: src/metadata_manager/metadata.rs [hunk 12]
```diff
@@ -145,12 +139,14 @@ pub struct Metadata {
     package_type_docs: DocumentStore,
     namespace_docs: DocumentStore,
     package_version_docs: DocumentStore,
-    untrusted_key_pair: SignatureKeyPair,
 }
 
-/// Methods that create metadata return this enum. In the case of a duplicate, the returned value
-/// includes the json of the duplicate metadata. It can be converted to a struct by passing it to
-/// `Signed::from_json_string()`
+pub fn now_as_iso8601_string() -> String {
```
Comment (prince-chrismc) [inline src/metadata_manager/metadata.rs:144]: I expected this to be in the test section since thats the only spot it's used 

https://github.com/pyrsia/pyrsia/blob/b5eb8a2a99224c0b7f10d2b48ad145bc1e3f56de/src/metadata_manager/metadata.rs#L368

File: src/metadata_manager/metadata.rs [hunk 13]
```diff
@@ -425,30 +386,25 @@ mod tests {
         let license_text_mimetype = LicenseTextMimeType::Text;
         let license_url = "https://example.com".to_string();
         let pv_metadata: serde_json::Map<String, Value> = serde_json::Map::new();
-        let creation_time = signed::now_as_iso8601_string();
-        let modified_time = signed::now_as_iso8601_string();
```
Comment (prince-chrismc) [inline src/metadata_manager/metadata.rs:429]: We should add these back like 

https://github.com/pyrsia/pyrsia/blob/b5eb8a2a99224c0b7f10d2b48ad145bc1e3f56de/src/metadata_manager/metadata.rs#L144

just remove the `pub`

File: src/node_manager/model/package_version.rs [hunk 14]
```diff
@@ -15,50 +15,98 @@
 */
 
 extern crate serde_json;
```
Comment (prince-chrismc) [inline src/node_manager/model/package_version.rs:17]: We have these if most of the files...

A few are missing from crates like this one... others have too many 🤢 not sure what to do about that

@tiainen @fishseabowl since you are the assigned reviewers what do you think?

Example of an extra
https://github.com/pyrsia/pyrsia/pull/494/files#diff-41604668c7de8413bd2cb0a539ef8d4253e530616a473a41733879334b1a7580R18

missing 
https://github.com/pyrsia/pyrsia/pull/494/files#diff-818bde2e9f654033515eecc77090bd5301af028216f43184821cc82ebbe9bfaaR17

File: src/node_manager/model/package_version.rs [hunk 15]
```diff
@@ -15,50 +15,98 @@
...
+            modified_time: Default::default(),
+            tags: Default::default(),
+        }
+    }
+
+    pub fn id(&self) -> &String {
+        &self.id
+    }
+
+    pub fn namespace_id(&self) -> &String {
```
Comment (prince-chrismc) [inline src/node_manager/model/package_version.rs:91]: Can these getters be deleted... that are not being used for the most part in the tests

**Review by prince-chrismc — COMMENTED** [review]
Lots of cleanup comments and questions for you but outstanding work!

File: Dockerfile (hunk 8 above)
[the remark this answers is not in the document]
Comment (tiainen) [reply Dockerfile:42]: A backslash + newline is treated as a line continuation. Currently, with the backslash, the command that is going to be executed will be the following:

```
apt-get install -y ca-certificates rm -rf /var/lib/apt/lists/*
```

This is not what we want (and will most likely even fail).

[the remark this answers is not in the document]
Comment (AbhijithGanesh) [reply Dockerfile:42]: I shall resolve this conversation

File: src/metadata_manager/metadata.rs (hunk 11 above)
Comment (AbhijithGanesh) [reply src/metadata_manager/metadata.rs:18]: Breaks otherwise.

File: src/metadata_manager/metadata.rs (hunk 12 above)
Comment (AbhijithGanesh) [reply src/metadata_manager/metadata.rs:144]: I shall move it

File: Cargo.toml (hunk 1 above)
Comment (AbhijithGanesh) [reply Cargo.toml:46]: Removed.

File: src/metadata_manager/metadata.rs (hunk 13 above)
Comment (AbhijithGanesh) [inline src/metadata_manager/metadata.rs:429]: This should be real values.

**Review by prince-chrismc — COMMENTED** [review]
Things to talk about at the standup

- usage of https://github.com/pyrsia/pyrsia/blob/21c43dcefd7da4e0791914e46b9dc64577a65f1a/src/node_manager/model/artifact.rs#L21

since it was mostly removed here (no idea it was magically being injected)

***

### Other code review comments

the `time` crate should be moved to dep deps

https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#development-dependencies

***

should be real values like before

https://github.com/pyrsia/pyrsia/pull/494/files#diff-3f1c9b8e3c1e709a2759abb3c53bc4b7cd6645e75f40c224679cb8240c49e520R400-R401

File: src/node_manager/model/namespace.rs [hunk 16]
```diff
@@ -16,28 +16,24 @@
...
-use signed_struct::signed_struct;
 
-#[signed_struct]
-#[derive(Debug, PartialEq)]
+#[derive(Debug, PartialEq, Deserialize, Serialize)]
 /// Describes a name space.
 pub struct Namespace {
     /// A uuid that uniquely identifies the name space
-    id: String,
+    pub id: String,
```
Comment (chb0github) [inline src/node_manager/model/namespace.rs:28]: Why the change to `pub`?

File: src/node_manager/model/package_version.rs (hunk 15 above)
Comment (chb0github) [reply src/node_manager/model/package_version.rs:91]: Well, I prefer getters to `pub` struct fields. But if the `pub` is needed for something else, then the getter methods are redundant

File: src/node_manager/model/namespace.rs (hunk 16 above)
Comment (AbhijithGanesh) [reply src/node_manager/model/namespace.rs:28]: Since we don't have the builder pattern, we need to expose the field to create it.

File: src/node_manager/model/package_version.rs (hunk 15 above)
Comment (AbhijithGanesh) [reply src/node_manager/model/package_version.rs:91]: It is being used as far as I know, its not redundant.

File: src/node_manager/model/package_version.rs (hunk 14 above)
Comment (tiainen) [reply src/node_manager/model/package_version.rs:17]: Declaring `extern crate` should [no longer be needed](https://doc.rust-lang.org/edition-guide/rust-2018/path-changes.html#no-more-extern-crate). Defining it in `Cargo.toml` is sufficient.

File: src/metadata_manager/metadata.rs (hunk 11 above)
Comment (tiainen) [reply src/metadata_manager/metadata.rs:18]: This is usually a sign that it's only used in tests. Move the `use` declaration inside the tests module.

File: src/node_manager/model/package_version.rs (hunk 14 above)
Comment (prince-chrismc) [reply src/node_manager/model/package_version.rs:17]: I did not even know that was a think 😱

File: Cargo.toml (hunk 9 above)
Comment (prince-chrismc) [reply Cargo.toml:63]: as below this should be a dev dep it's only needed for our tests

File: src/metadata_manager/metadata.rs (hunk 11 above)
Comment (prince-chrismc) [reply src/metadata_manager/metadata.rs]: you hunch was spot on https://github.com/pyrsia/pyrsia/pull/494/commits/12bf7aad83565fa8c62178b25d5981eca34a36ea

File: src/node_manager/model/package_version.rs (hunk 15 above)
Comment (prince-chrismc) [reply src/node_manager/model/package_version.rs:91]: The code should be cleaned up, they are required

File: src/metadata_manager/metadata.rs [hunk 17]
```diff
@@ -71,21 +70,15 @@ fn ix_package_types() -> Vec<IndexSpec> {
...
-        .expect("Failed to create package type struct for pre-installation");
-    package_type.sign_json(
-        JwsSignatureAlgorithms::RS512,
-        &key_pair.private_key,
-        &key_pair.public_key,
-    )?;
-    Ok(vec![package_type.json().unwrap_or_else(|| {
-        "package type for pre-installation somehow does not have JSON".to_string()
-    })])
+fn init_package_types() -> Result<Vec<String>> {
```
Comment (fishseabowl) [inline src/metadata_manager/metadata.rs:73]: I'm curious if we don't support signed packages or if we still support signed packages just remove Openssl dependency.

Comment (AbhijithGanesh) [reply src/metadata_manager/metadata.rs:73]: We don't support signing, I am not sure if that subsequently means we don't support signed packages. We don't support signing off packages (by pyrsia)

File: src/metadata_manager/metadata.rs [hunk 18]
```diff
@@ -294,30 +272,21 @@ fn fetch_package_version(
...
-            Err(error) => bail!(
-                "Failed to create package_type record: {:?}\nError is {}",
-                signed,
-                error.to_string()
-            ),
-        },
-        None => bail!(
-            "A supposedly trusted metadata struct is missing its JSON: {:?}",
-            signed
+fn insert_metadata(ds: &DocumentStore, signed: String) -> anyhow::Result<MetadataCreationStatus> {
```
Comment (tiainen) [inline src/metadata_manager/metadata.rs:280]: The name `signed` no longer correctly reflects its meaning. Just `metadata` is probably better.

File: src/docker/v2/handlers/manifests.rs [hunk 19]
```diff
@@ -200,15 +199,8 @@ fn internal_error_response(
 fn sign_and_save_package_version(
```
Comment (tiainen) [inline src/docker/v2/handlers/manifests.rs:199]: This method is no longer signing:

```suggestion
fn save_package_version(
```

File: Cargo.toml (hunk 9 above)
Comment (prince-chrismc) [reply Cargo.toml:63]: bump

File: src/docker/v2/handlers/manifests.rs [hunk 20]
```diff
@@ -945,47 +936,47 @@ mod tests {
         assert_eq!("hello-world", package_version.name());
         assert_eq!(PackageTypeName::Docker, *package_version.pkg_type());
         assert_eq!("v3.1", package_version.version());
```
Comment (prince-chrismc) [inline src/docker/v2/handlers/manifests.rs:938]: These are pub

```suggestion
        assert_eq!("hello-world", package_version.name);
        assert_eq!(PackageTypeName::Docker, *package_version.pkg_type);
        assert_eq!("v3.1", package_version.version);
```

File: src/metadata_manager/metadata.rs [hunk 21]
```diff
@@ -358,24 +320,18 @@ mod tests {
...
-            &key_pair.public_key,
-        )?;
+        let namespace_clone = path.clone();
+        let namespace = Namespace {
+            id: id,
+            package_type: PackageTypeName::Docker,
+            namespace_path: namespace_clone,
+            administrators: vec![vec![]],
+            creation_time: Some(String::from("")),
+            modified_time: Some(String::from("")),
```
Comment (prince-chrismc) [inline src/metadata_manager/metadata.rs:333]: These should be `timestamp` like before

File: src/node_manager/model/package_version.rs (hunk 15 above)
Comment (prince-chrismc) [reply src/node_manager/model/package_version.rs:91]: bump

File: src/metadata_manager/metadata.rs (hunk 18 above)
Comment (prince-chrismc) [reply src/metadata_manager/metadata.rs:280]:
```suggestion
fn insert_metadata(ds: &DocumentStore, metadata: String) -> anyhow::Result<MetadataCreationStatus> {
```

File: src/metadata_manager/metadata.rs (hunk 21 above)
Comment (AbhijithGanesh) [reply src/metadata_manager/metadata.rs:333]: got it

**Review by prince-chrismc — COMMENTED** [review]
Some outstanding comments to address

- https://github.com/pyrsia/pyrsia/pull/494/files#r829390265
- https://github.com/pyrsia/pyrsia/pull/494/files#r829204456

File: src/node_manager/model/package_version.rs (hunk 15 above)
Comment (AbhijithGanesh) [reply src/node_manager/model/package_version.rs:91]: These cannot be deleted.

**Review by prince-chrismc — APPROVED** [review]
LGTM!

**Review by fishseabowl — COMMENTED** [review]
if possible, please make struct fields private.

**Comment by AbhijithGanesh** [comment]
@fishseabowl Can you please give me the file name, GitHub doesn't show me the changes (because the diff is too large) . Thanks : )

**Comment by fishseabowl** [comment]
> @fishseabowl Can you please give me the file name, GitHub doesn't show me the changes (because the diff is too large) . Thanks :)

@AbhijithGanesh In Rust, the visibility of struct fields defaults to private. Keep it if possible.  Files list: src/node_manager/model/package_type.rs, src/node_manager/model/package_version.rs,  and src/node_manager/model/namespace.rs. Thanks

File: .github/workflows/rust.yml [hunk 22]
```diff
@@ -186,4 +181,4 @@ jobs:
         context: .
         push: true
         file: installers/docker/CodeCoverage.Dockerfile
-        tags: pyrsiaoss/codecoverage:1.0
+        tags: pyrsiaoss/codecoverage:1.0
```
Comment (prince-chrismc) [inline .github/workflows/rust.yml:184]: Should be a newline here

File: Cargo.toml [hunk 23]
```diff
@@ -116,4 +113,4 @@ lto = false
 panic = 'unwind'
 incremental = false
 codegen-units = 16
-rpath = false
+rpath = false
```
Comment (prince-chrismc) [inline Cargo.toml:116]: new line

File: src/node_manager/model/namespace.rs [hunk 24]
```diff
@@ -1,12 +1,9 @@
 /*
    Copyright 2021 JFrog Ltd
-
```
Comment (prince-chrismc) [inline src/node_manager/model/namespace.rs:3]: Unexpected. this should fail CI

**Review by prince-chrismc — COMMENTED** [review]
Super minor nits,

seems like your last commit was during the service disruptions so new commit should trigger CI

**Comment by AbhijithGanesh** [comment]
Can you trigger it manually please?

File: .github/workflows/rust.yml [hunk 25]
```diff
@@ -1,184 +1,184 @@
 name: Rust
 
 on:
-  push:
-    branches: [ main ]
-  pull_request:
+    push:
```
Comment (prince-chrismc) [inline .github/workflows/rust.yml:4]: these changes look out of place 🤔

Comment (AbhijithGanesh) [reply .github/workflows/rust.yml:4]: How should the file be ? What change do I have to do?

**Comment by AbhijithGanesh** [comment]
@prince-chrismc @fishseabowl please review

File: .github/workflows/rust.yml (hunk 22 above)
Comment (fishseabowl) [reply .github/workflows/rust.yml:184]:
```suggestion
tags: pyrsiaoss/codecoverage:1.0
```

File: src/blockchain/Cargo.toml [hunk 26]
```diff
@@ -27,4 +27,4 @@ name = "simple_node"
 
 [dev-dependencies]
 pretty_env_logger = "0.4.0"
-tokio = { version = "1.17.0", features = [ "macros", "rt-multi-thread", "io-std" ] }
+tokio = { version = "1.17.0", features = [ "macros", "rt-multi-thread", "io-std" ] }
```
Comment (fishseabowl) [inline src/blockchain/Cargo.toml:30]: Should be a newline here
```suggestion
tokio = { version = "1.17.0", features = [ "macros", "rt-multi-thread", "io-std" ] }

```

**Review by fishseabowl — COMMENTED** [review]
namespaces.db, package_types.db and package_versions.db should not be merged into codebase

**Comment by AbhijithGanesh** [comment]
I ll remove them

File: .github/workflows/rust.yml [hunk 27]
```diff
@@ -186,4 +181,4 @@ jobs:
         context: .
         push: true
         file: installers/docker/CodeCoverage.Dockerfile
-        tags: pyrsiaoss/codecoverage:1.0
+tags: pyrsiaoss/codecoverage:1.0
```
Comment (prince-chrismc) [inline .github/workflows/rust.yml:184]: This looks incorrect 🤔 this file does not need to be changed

**Review by prince-chrismc — APPROVED** [review]

**Review by fishseabowl — APPROVED** [review]
LGTM
